docs/decisions/0010_audit_logging.rst

```diff
@@ -0,0 +1,36 @@
+10. Audit logging
+#################
+
+Status
+******
+
+- **Accepted** 2023-01-31
+
+Context
+*******
+
+We do not currently have a reliable source for information on the timestamps and processing time of Kafka messages. This information is necessary for a variety of purposes:
+
+- Determining what timestamp to start replaying messages from, if a consumer or producer has misbehaved
+- Investigating whether we have messages that are taking an excessively long time to process
+- Checking the logs for what happened right *before* an error
+- And just diagnosing bugs, in general
+
+The Kafka broker may have information about when the broker received the messages, but probably will not record when messages were consumed, or how long it took the consumer to process a message. New Relic has some information, but due to sampling we don't have a complete picture. (Usually we record information for 99+% of events, but over an example 1 month window we see average capture drop as low as 70%. This may be due to heavier loss during replays.) Our logs also have some information, but only for errors.
+
+Decision
+********
+
+We will log at least minimal information about all Kafka messages that we produce and consume. These log messages will not necessarily have sufficient information to reconstruct a message, but will identify messages uniquely (topic, partition, and offset) and include timestamps that are necessary for reconstructing a timeline of activity. Beyond that we will probably just include identifying information such as message keys and unique IDs, as they are useful for searches but should not add too much chaff to the logs.
+
+Timestamps will be recorded implicitly, via the logging framework's own log formatting.
+
+Consequences
+************
+
+Log volume will be increased, likely not by too much.
```

Comment:
Thoughts on adding a toggle to enable/disable audit logging?

Reply:
I suppose it might make sense, but I'd be inclined to hold off until it seems necessary. Were you thinking default-on or default-off?

Reply:
I was thinking default-on, but either way. Just a kill switch if needed.

Reply:
Toggle added.